src/librbd/api/Mirror.cc

```diff
@@ -2098,6 +2345,1689 @@ void Mirror<I>::image_snapshot_create(I *ictx, uint32_t flags,
+    return r;
+  }
+
+  cls::rbd::MirrorGroup mirror_group;
+  r = cls_client::mirror_group_get(&group_ioctx, group_id, &mirror_group);
+  if (r == -ENOENT) {
+    ldout(cct, 10) << "mirroring for group " << group_name
+                   << " disabled" << dendl;
+    return -EINVAL;
+  } else if (r < 0) {
+    lderr(cct) << "failed to retrieve mirror group metadata: "
+               << cpp_strerror(r) << dendl;
+    return r;
+  } else if (mirror_group.mirror_image_mode !=
+             cls::rbd::MIRROR_IMAGE_MODE_SNAPSHOT) {
+    return -EOPNOTSUPP;
+  }
+
+  cls::rbd::MirrorSnapshotState state;
+  r = get_last_mirror_snapshot_state(group_ioctx, group_id, &state);
+  if (r == -ENOENT) {
+    state = cls::rbd::MIRROR_SNAPSHOT_STATE_PRIMARY;
+    r = 0;
+  }
+  if (r < 0) {
+    return r;
+  }
+
+  if (state != cls::rbd::MIRROR_SNAPSHOT_STATE_PRIMARY) {
+    lderr(cct) << "group " << group_name << " is not primary" << dendl;
+    return -EINVAL;
+  }
+
+  std::string group_snap_id = librbd::util::generate_uuid(group_ioctx);
+  cls::rbd::GroupSnapshot group_snap{
+      group_snap_id,
+      cls::rbd::MirrorGroupSnapshotNamespace{},
+      calc_ind_mirror_snap_name(group_ioctx.get_id(), group_id, group_snap_id),
+      cls::rbd::GROUP_SNAPSHOT_STATE_INCOMPLETE};
+
+  std::vector<uint64_t> quiesce_requests;
+  std::vector<I *> image_ctxs;
+  r = prepare_group_images(group_ioctx, group_id, &image_ctxs,
+                           &group_snap, quiesce_requests,
+                           cls::rbd::MIRROR_SNAPSHOT_STATE_PRIMARY_DEMOTED,
+                           flags);
+  if (r != 0) {
+    return r;
+  }
+
+  int ret_code = 0;
+  std::vector<uint64_t> snap_ids(image_ctxs.size(), CEPH_NOSNAP);
+  std::vector<C_SaferCond*> on_finishes(image_ctxs.size(), nullptr);
+  C_SaferCond* on_finish;
+
+  for (size_t i = 0; i < image_ctxs.size(); i++) {
+    C_SaferCond* on_finish = new C_SaferCond;
+    image_demote(image_ctxs[i], group_ioctx.get_id(), group_id, group_snap_id,
+            &snap_ids[i], on_finish);
+    on_finishes[i] = on_finish;
+  }
+
+  for (size_t i = 0; i < image_ctxs.size(); i++) {
+    r = 0;
+    if (on_finishes[i]) {
+      r = on_finishes[i]->wait();
+      delete on_finishes[i];
+    }
+    if (r < 0) {
+      if (ret_code == 0) {
+        ret_code = r;
+      }
+    } else {
+      group_snap.snaps[i].snap_id = snap_ids[i];
+    }
+  }
+
+  std::string group_header_oid = librbd::util::group_header_name(group_id);
+  if (ret_code < 0) {
+    // undo
+    ldout(cct, 20) << "undoing group demote: " << ret_code << dendl;
+    remove_interim_snapshots(group_ioctx, group_header_oid, &image_ctxs, &group_snap);
+    std::fill(snap_ids.begin(), snap_ids.end(), CEPH_NOSNAP);
+    group_snap.snaps.clear();
+    if (!quiesce_requests.empty()) {
+      util::notify_unquiesce(image_ctxs, quiesce_requests);
+    }
+    close_images(&image_ctxs);
+
+    r = prepare_group_images(group_ioctx, group_id, &image_ctxs,
```

Comment:
This would reopen all the images that were closed 2 lines above again, right? Is there a better way?